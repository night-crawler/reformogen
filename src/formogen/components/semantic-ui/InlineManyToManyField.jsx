import _ from 'lodash';
import PropTypes from 'prop-types';
import React from 'react';
import Select from 'react-select';
import { Form } from 'semantic-ui-react';
import { extractIdentity } from '../../utils';

import { errorsType, layoutOptsType } from '../fieldPropTypes';

import Label from './Label';
import { MessageList } from './MiscComponents';


makeOptions.propTypes = PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
}));

function makeOptions(fieldData) {
    return fieldData.map(({ id, name }) => ( { label: name, value: id } ));
}


InlineManyToManyField.propTypes = {
    type: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    value: PropTypes.arrayOf(PropTypes.any),
    data: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
    })),
    max_length: PropTypes.number,
    help_text: PropTypes.string.isRequired,
    placeholder: PropTypes.string,
    errors: errorsType,

    required: PropTypes.bool,
    editable: PropTypes.bool,

    helpTextOnHover: PropTypes.bool,
    layoutOpts: layoutOptsType,

    updateProps: PropTypes.func,
    onChange: PropTypes.func,
};
export default function InlineManyToManyField(props) {
    const handleChange = (val) => {
        const plainIds = val.map(({ value }) => ( value * 1 ));
        props.onChange(null, { name: props.name, value: plainIds });
    };

    let _props = {
        closeOnSelect: false,
        disabled: !props.editable,
        multi: true,
        onChange: handleChange,
        options: makeOptions(props.data),
        placeholder: props.placeholder,
        // simpleValue: true,
        value: extractIdentity(props.value),
        inputProps: { type: 'react-type' },  // fixes broken semantic markup
        removeSelected: true,
        // rtl: this.state.rtl,
    };

    if (_.isFunction(props.updateProps)) {
        _props = props.updateProps(_props, props);
    }

    return (
        <Form.Field
            required={ props.required }
            disabled={ !props.editable }
            width={ props.layoutOpts.width }
            error={ !_.isEmpty(props.errors) }
        >
            <Label { ...props } />
            <Select { ..._props } />

            { !props.helpTextOnHover ? <span className='help-text'>{ props.help_text }</span> : '' }
            <MessageList messages={ props.errors } />
        </Form.Field>
    );
}
