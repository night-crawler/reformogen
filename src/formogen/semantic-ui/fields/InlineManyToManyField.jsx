import React from 'react';
import PropTypes from 'prop-types';
import { Form } from 'semantic-ui-react';
import Select from 'react-select';
import { map, isEmpty } from 'lodash';

import { errorsType, layoutOptsType } from '../../fieldPropTypes';
import { FieldLabel } from '../common/FieldLabel';
import { ErrorsList } from '../common/ErrorsList';


InlineManyToManyField.propTypes = {
  type: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.arrayOf(PropTypes.any),
  data: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
  })),
  max_length: PropTypes.number,
  help_text: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  errors: errorsType,

  required: PropTypes.bool,
  editable: PropTypes.bool,

  helpTextOnHover: PropTypes.bool,
  layoutOpts: layoutOptsType,

  getOptionLabel: PropTypes.func,
  getOptionValue: PropTypes.func,
  closeOnSelect: PropTypes.bool,

  onChange: PropTypes.func,
};
InlineManyToManyField.defaultProps = {
  getOptionLabel: ({ name }) => name,
  getOptionValue: ({ id }) => id,
  closeOnSelect: true,
};
export function InlineManyToManyField(props) {
  const handleChange = val =>
    props.onChange(null, { name: props.name, value: map(val, 'id') });

  return (
    <Form.Field
      required={ props.required }
      disabled={ !props.editable }
      width={ props.layoutOpts.width }
      error={ !isEmpty(props.errors) }
    >
      <FieldLabel { ...props } />
      <Select 
        isClearable={ props.editable }
        closeOnSelect={ props.closeOnSelect }
        isDisabled={ !props.editable }
        isMulti={ true }
        onChange={ handleChange }
        options={ props.data }
        placeholder={ props.placeholder }
        value={ props.data.filter(({ id }) => props.value.includes(id) ) }

        getOptionLabel={ props.getOptionLabel }
        getOptionValue={ props.getOptionValue }
      />

      { !props.helpTextOnHover 
        ? <span className='help-text'>{ props.help_text }</span> 
        : '' 
      }
      <ErrorsList messages={ props.errors } />
    </Form.Field>
  );
}
