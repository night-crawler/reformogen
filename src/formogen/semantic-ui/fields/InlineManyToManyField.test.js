import { InlineManyToManyField } from './InlineManyToManyField';


it('<InlineManyToManyField />', () => {
  const wrapper = shallow(
    <InlineManyToManyField
      type='InlineManyToManyField'
      name='InlineManyToManyField: name'
      verbose_name='InlineManyToManyField: verbose_name'
      help_text='InlineManyToManyField: help_text'
      layoutOpts={ { width: 1 } }
    />
  );
  expect(wrapper).toMatchSnapshot();
});
