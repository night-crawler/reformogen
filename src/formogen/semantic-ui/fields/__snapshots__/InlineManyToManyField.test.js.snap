// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InlineManyToManyField /> 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <InlineManyToManyField
    closeOnSelect={true}
    data={Array []}
    displayOptions={
      Object {
        "width": 1,
      }
    }
    getOptionLabel={[Function]}
    getOptionValue={[Function]}
    help_text="InlineManyToManyField: help_text"
    name="InlineManyToManyField: name"
    type="InlineManyToManyField"
    value={Array []}
    verbose_name="InlineManyToManyField: verbose_name"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <FieldLabel
          closeOnSelect={true}
          data={Array []}
          displayOptions={
            Object {
              "width": 1,
            }
          }
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          help_text="InlineManyToManyField: help_text"
          name="InlineManyToManyField: name"
          type="InlineManyToManyField"
          value={Array []}
          verbose_name="InlineManyToManyField: verbose_name"
        />,
        <StateManager
          closeOnSelect={true}
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          isDisabled={true}
          isMulti={true}
          onChange={[Function]}
          options={Array []}
          value={Array []}
        />,
        <span
          className="help-text"
        >
          InlineManyToManyField: help_text
        </span>,
        <ErrorsList
          color="red"
          header="Errors"
        />,
      ],
      "disabled": true,
      "error": false,
      "required": undefined,
      "width": 1,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "closeOnSelect": true,
          "data": Array [],
          "displayOptions": Object {
            "width": 1,
          },
          "getOptionLabel": [Function],
          "getOptionValue": [Function],
          "help_text": "InlineManyToManyField: help_text",
          "name": "InlineManyToManyField: name",
          "type": "InlineManyToManyField",
          "value": Array [],
          "verbose_name": "InlineManyToManyField: verbose_name",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "closeOnSelect": true,
          "defaultInputValue": "",
          "defaultMenuIsOpen": false,
          "defaultValue": null,
          "getOptionLabel": [Function],
          "getOptionValue": [Function],
          "isClearable": undefined,
          "isDisabled": true,
          "isMulti": true,
          "onChange": [Function],
          "options": Array [],
          "placeholder": undefined,
          "value": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "InlineManyToManyField: help_text",
          "className": "help-text",
        },
        "ref": null,
        "rendered": "InlineManyToManyField: help_text",
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "color": "red",
          "header": "Errors",
          "messages": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <FieldLabel
            closeOnSelect={true}
            data={Array []}
            displayOptions={
              Object {
                "width": 1,
              }
            }
            getOptionLabel={[Function]}
            getOptionValue={[Function]}
            help_text="InlineManyToManyField: help_text"
            name="InlineManyToManyField: name"
            type="InlineManyToManyField"
            value={Array []}
            verbose_name="InlineManyToManyField: verbose_name"
          />,
          <StateManager
            closeOnSelect={true}
            defaultInputValue=""
            defaultMenuIsOpen={false}
            defaultValue={null}
            getOptionLabel={[Function]}
            getOptionValue={[Function]}
            isDisabled={true}
            isMulti={true}
            onChange={[Function]}
            options={Array []}
            value={Array []}
          />,
          <span
            className="help-text"
          >
            InlineManyToManyField: help_text
          </span>,
          <ErrorsList
            color="red"
            header="Errors"
          />,
        ],
        "disabled": true,
        "error": false,
        "required": undefined,
        "width": 1,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "closeOnSelect": true,
            "data": Array [],
            "displayOptions": Object {
              "width": 1,
            },
            "getOptionLabel": [Function],
            "getOptionValue": [Function],
            "help_text": "InlineManyToManyField: help_text",
            "name": "InlineManyToManyField: name",
            "type": "InlineManyToManyField",
            "value": Array [],
            "verbose_name": "InlineManyToManyField: verbose_name",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "closeOnSelect": true,
            "defaultInputValue": "",
            "defaultMenuIsOpen": false,
            "defaultValue": null,
            "getOptionLabel": [Function],
            "getOptionValue": [Function],
            "isClearable": undefined,
            "isDisabled": true,
            "isMulti": true,
            "onChange": [Function],
            "options": Array [],
            "placeholder": undefined,
            "value": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "InlineManyToManyField: help_text",
            "className": "help-text",
          },
          "ref": null,
          "rendered": "InlineManyToManyField: help_text",
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "color": "red",
            "header": "Errors",
            "messages": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
